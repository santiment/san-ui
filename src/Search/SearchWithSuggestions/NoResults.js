import React from 'react'
import styles from './NoResults.module.scss'
import commonStyles from './SuggestionItem.module.scss'

const NoResults = ({ isSearching }) => (
  <div className={commonStyles.suggestion + ' ' + styles.noresults}>
    {!isSearching ? 'No results found' : 'Searching...'}
  </div>
)

export default NoResults
