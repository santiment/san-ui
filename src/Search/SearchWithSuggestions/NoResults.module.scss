@import '../../mixins';

.noresults {
  @include text-normal;

  cursor: initial;
  min-height: auto;
  text-align: center;

  &:hover,
  &:focus {
    background-color: inherit !important;
  }
}
