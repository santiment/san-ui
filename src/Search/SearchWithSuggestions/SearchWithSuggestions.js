import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import Suggestions from './Suggestions'
import Panel from '../../Panel'
import Search from '../Search'
import { flatCategories } from './utils'
import styles from './SearchWithSuggestions.module.scss'

export const SUGGESTION_MORE = 'SUGGESTION_MORE'
export const SUGGESTION_MORE_ITEM = 'SUGGESTION_MORE_ITEM'
export const MORE = {
  category: SUGGESTION_MORE,
  item: SUGGESTION_MORE_ITEM
}

let debounceTimer
const debounce = (clb, time) => clbArgs => {
  clearTimeout(debounceTimer)
  debounceTimer = setTimeout(() => clb(clbArgs), time)
}

class SearchWithSuggestions extends PureComponent {
  static propTypes = {
    data: PropTypes.arrayOf(
      PropTypes.shape({
        title: PropTypes.string.isRequired,
        items: PropTypes.array.isRequired,
        suggestionContent: PropTypes.func.isRequired,
        predicate: PropTypes.func.isRequired,
        sorter: PropTypes.func,
        maxSuggestions: PropTypes.number
      })
    ).isRequired,
    onSuggestionSelect: PropTypes.func,
    onSuggestionsUpdate: PropTypes.func,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    maxSuggestions: PropTypes.number,
    debounceTime: PropTypes.number,
    inputProps: PropTypes.object,
    suggestionsProps: PropTypes.object,
    dontResetStateAfterSelection: PropTypes.bool,
    className: PropTypes.string,
    emptySuggestions: PropTypes.array,
    withMoreSuggestions: PropTypes.bool,
    header: PropTypes.element,
    multipleCheck: PropTypes.bool
  }

  static defaultProps = {
    maxSuggestions: 5,
    onSuggestionSelect: () => {},
    onSuggestionsUpdate: () => {},
    inputProps: {},
    suggestionsProps: {},
    debounceTime: 200,
    dontResetStateAfterSelection: false,
    withMoreSuggestions: true,
    header: null,
    multipleCheck: false,
    value: '',
    defaultValue: '',
    className: ''
  }

  static getDerivedStateFromProps (
    { value, emptySuggestions },
    { searchTerm, suggestedCategories, isSearching }
  ) {
    if (
      !searchTerm &&
      !isSearching &&
      emptySuggestions !== suggestedCategories
    ) {
      const checkedEmptySuggestions = emptySuggestions || []
      return {
        suggestions: flatCategories(checkedEmptySuggestions, []),
        suggestedCategories: checkedEmptySuggestions
      }
    }

    return null
  }

  searchRef = React.createRef()

  state = {
    suggestions: [],
    searchTerm: this.props.defaultValue,
    isFocused: false,
    cursor: 0,
    isSearching: false,
    onClear: () => {}
  }

  componentWillUnmount () {
    clearTimeout(debounceTimer)
    window.removeEventListener('keydown', this.onKeyDown)
    document.removeEventListener('click', this.handleClickAway)
  }

  onInputChange = searchTerm => {
    this.setState(
      prevState => ({
        ...prevState,
        searchTerm,
        isSearching: true
      }),
      this.filterData
    )
  }

  onSuggestionSelect = suggestion => {
    const {
      dontResetStateAfterSelection,
      onSuggestionSelect,
      emptySuggestions
    } = this.props

    this.onBlur()
    this.setState(
      dontResetStateAfterSelection
        ? {
            isSearching: false
          }
        : {
            isSearching: false,
            searchTerm: '',
            suggestions: emptySuggestions
              ? flatCategories(emptySuggestions, [])
              : [],
            suggestedCategories: emptySuggestions || [],
            cursor: 0
          },
      () => onSuggestionSelect(suggestion)
    )
  }

  filterData = debounce(() => {
    const {
      data,
      emptySuggestions = [],
      onSuggestionsUpdate,
      withMoreSuggestions,
      maxSuggestions: commonMaxSuggestions
    } = this.props

    this.setState(
      prevState => {
        const suggestedCategories = prevState.searchTerm
          ? data
              .map(
                ({
                  items,
                  predicate,
                  sorter,
                  maxSuggestions = commonMaxSuggestions,
                  ...rest
                }) => {
                  const filteredItems = items.filter(
                    predicate(prevState.searchTerm)
                  )

                  if (sorter) {
                    filteredItems.sort(sorter(prevState.searchTerm))
                  }

                  return {
                    ...rest,
                    items: filteredItems.slice(0, maxSuggestions)
                  }
                }
              )
              .filter(({ items }) => items.length)
          : emptySuggestions

        const suggestions = flatCategories(
          suggestedCategories,
          withMoreSuggestions && prevState.searchTerm ? [MORE] : []
        )

        const cursor = +Boolean(withMoreSuggestions && suggestions.length)

        return {
          ...prevState,
          suggestedCategories,
          suggestions,
          cursor,
          isSearching: false
        }
      },
      () => onSuggestionsUpdate(this.state.suggestions)
    )
  }, this.props.debounceTime)

  handleClickAway = ({ target }) => {
    if (!this.searchRef.current.contains(target)) {
      this.onBlur()
    }
  }

  onFocus = () => {
    document.addEventListener('click', this.handleClickAway)
    window.addEventListener('keydown', this.onKeyDown)
    this.setState({ isFocused: true }, this.props.onFocus)
  }

  onBlur = () => {
    document.removeEventListener('click', this.handleClickAway)
    window.removeEventListener('keydown', this.onKeyDown)
    this.setState({ isFocused: false }, this.props.onBlur)
  }

  onKeyDown = evt => {
    const { suggestions, cursor } = this.state
    const { key } = evt

    let newCursor = cursor
    let selectedSuggestion

    switch (key) {
      case 'ArrowUp':
        evt.preventDefault()
        newCursor = cursor - 1
        break
      case 'ArrowDown':
        evt.preventDefault()
        newCursor = cursor + 1
        break
      case 'Enter':
        selectedSuggestion = suggestions[cursor]
        if (selectedSuggestion) this.onSuggestionSelect(selectedSuggestion)
        this.state.onClear()
        this.onFocus()
        return
      default:
        return
    }

    const maxCursor = suggestions.length

    newCursor = newCursor % maxCursor

    const nextCursor = newCursor < 0 ? maxCursor - 1 : newCursor
    this.setState({ cursor: nextCursor })
  }

  render () {
    const {
      suggestedCategories = [],
      searchTerm,
      isFocused,
      isSearching,
      cursor,
      suggestions
    } = this.state
    const {
      inputProps = {},
      suggestionsProps = {},
      className,
      emptySuggestions,
      withMoreSuggestions,
      header
    } = this.props
    const cursorSuggestion = suggestions[cursor]
    return (
      <Search
        className={className}
        onFocus={this.onFocus}
        onChange={this.onInputChange}
        wrapperRef={this.searchRef}
        setOnClear={onClearFn =>
          this.setState(prevState => ({
            ...prevState,
            onClear: onClearFn
          }))
        }
        {...inputProps}
      >
        {isFocused && (emptySuggestions || searchTerm) && (
          <Panel
            variant='modal'
            className={styles.suggestions}
            {...suggestionsProps}
          >
            {header}
            <Suggestions
              cursorItem={cursorSuggestion && cursorSuggestion.item}
              suggestedCategories={suggestedCategories}
              isSearching={isSearching}
              searchTerm={searchTerm}
              withMoreSuggestions={withMoreSuggestions}
              onSuggestionSelect={this.onSuggestionSelect}
            />
          </Panel>
        )}
      </Search>
    )
  }
}

export default SearchWithSuggestions
