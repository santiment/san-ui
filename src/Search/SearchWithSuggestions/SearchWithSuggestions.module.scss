@import '../../mixins';

$border-color-night: #4a4a4a;

.wrapper {
  position: relative;
}

.suggestions {
  position: absolute;
  top: 100%;
  width: 100%;
  margin-top: 4px;
  z-index: 5;
  overflow: auto;
  max-height: 180px;
}

.groupSuggestions {
  max-height: 420px;
}

.divider {
  border-bottom: 1px solid var(--porcelain);
  margin-bottom: 14px;
}

.viewAllResults {
  cursor: pointer;
  margin: 0 -8px 14px -8px;
  padding: 14px 0 14px 24px;
  color: var(--mirage);

  @include text('body-3');

  &:hover {
    color: var(--jungle-green);
  }
}
