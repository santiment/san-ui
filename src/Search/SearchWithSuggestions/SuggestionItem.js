import React, { useEffect, useRef } from 'react'
import cx from 'classnames'
import Button from '../../Button'
import styles from './SuggestionItem.module.scss'

const SuggestionItem = ({
  isActive,
  onSuggestionSelect,
  suggestionContent,
  suggestion
}) => {
  const myRef = useRef(null)

  useEffect(() => {
    isActive && myRef && myRef.current && myRef.current.scrollIntoView(false)
  }, [isActive, myRef])

  return (
    <Button
      fluid
      forwardedRef={isActive ? myRef : undefined}
      variant='ghost'
      className={cx(styles.suggestion, isActive && styles.cursored)}
      onMouseDown={() => onSuggestionSelect(suggestion)}
    >
      {suggestionContent(suggestion)}
    </Button>
  )
}

export default SuggestionItem
