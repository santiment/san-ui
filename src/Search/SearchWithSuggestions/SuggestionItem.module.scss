@import '../../mixins';

.suggestion {
  padding: 6px 8px;
  color: var(--mirage);
  height: auto;
  min-height: 32px;
  text-align: left;

  @include text('body-3');

  &:hover,
  &:focus {
    color: var(--jungle-green);
  }
}

.cursored {
  background: var(--athens);
}
