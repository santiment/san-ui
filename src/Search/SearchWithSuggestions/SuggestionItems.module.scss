@import '../../mixins';

.groupLabel {
  @include text('caption');

  margin: 0 0 10px 8px;
  display: flex;
  align-items: center;
  color: var(--waterloo);
  font-weight: 600;
}
