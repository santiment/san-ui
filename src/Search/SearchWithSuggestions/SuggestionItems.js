import React, { Fragment } from 'react'
import SuggestionItemsList from './SuggestionsItemsList'
import ViewAllResults from './ViewAllResults'
import { getLengthOfSuggestions, isGroups } from './SearchWithSuggestions'
import NoResults from './NoResults'
import commonStyles from './SearchWithSuggestions.module.scss'
import styles from './SuggestionItems.module.scss'

const SuggestionItems = ({
  suggestions,
  cursor,
  onSuggestionSelect,
  suggestionContent,
  isSearching,
  maxSuggestions,
  onViewAllResults,
  searchTerm
}) => {
  const getSliced = data =>
    maxSuggestions ? data.slice(0, maxSuggestions) : data

  if (isGroups(suggestions)) {
    const noData = getLengthOfSuggestions(suggestions, maxSuggestions) === 0
    const types = Object.keys(suggestions)

    let fromCounter = 0

    return (
      <>
        {noData ? (
          <NoResults isSearching={isSearching} />
        ) : (
          <>
            <ViewAllResults
              searchTerm={searchTerm}
              onViewAllResults={onViewAllResults}
              suggestions={suggestions}
            />
            {types.map((key, index) => {
              const { label, options } = suggestions[key]

              const formattedOptions = getSliced(options)

              if (!formattedOptions.length) {
                return null
              }

              fromCounter += formattedOptions.length

              return (
                <div
                  key={key}
                  className={
                    index !== types.length - 1
                      ? commonStyles.divider
                      : undefined
                  }
                >
                  {label && <div className={styles.groupLabel}>{label}</div>}
                  <SuggestionItemsList
                    fromCounter={fromCounter - formattedOptions.length}
                    suggestions={formattedOptions}
                    cursor={cursor}
                    onSuggestionSelect={selected =>
                      onSuggestionSelect([key, selected])
                    }
                    suggestionContent={suggestionContent}
                  />
                </div>
              )
            })}
          </>
        )}
      </>
    )
  } else {
    const sliced = getSliced(suggestions)
    return sliced.length > 0 ? (
      <>
        <ViewAllResults
          searchTerm={searchTerm}
          onViewAllResults={onViewAllResults}
          suggestions={suggestions}
        />
        <SuggestionItemsList
          suggestions={sliced}
          cursor={cursor}
          onSuggestionSelect={onSuggestionSelect}
          suggestionContent={suggestionContent}
        />
      </>
    ) : (
      <NoResults isSearching={isSearching} />
    )
  }
}

export default SuggestionItems
