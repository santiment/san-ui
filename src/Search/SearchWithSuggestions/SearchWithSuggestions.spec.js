import React from 'react'
import { shallow, mount } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import SearchWithSuggestions from './SearchWithSuggestions'

const assets = [
  'Bibox Token',
  'Bigbom',
  'Binance Coin',
  'BioCoin',
  'BitBay',
  'bitcoin'
]

const simplePredicate = searchTerm => item =>
  item.toUpperCase().includes(searchTerm.toUpperCase())

const data = [
  {
    title: 'Assets',
    items: assets,
    predicate: simplePredicate,
    suggestionContent: suggestion => suggestion
  }
]

describe('SearchWithSuggestions', () => {
  it('should render correctly', () => {
    const output = shallow(
      <SearchWithSuggestions
        data={data}
        onSuggestionSelect={() => console.log('check')}
        maxSuggestions={5}
      />
    )
    expect(shallowToJson(output)).toMatchSnapshot()
  })

  it('should render Bigbom as a child', () => {
    const output = mount(
      <SearchWithSuggestions
        data={data}
        onSuggestionSelect={() => console.log('check')}
        maxSuggestions={5}
      />
    )
    const searchInput = output.find('Input')
    searchInput.children().instance().value = 'Big'
    searchInput.children().simulate('change')
    expect(shallowToJson(output)).toMatchSnapshot()
  })
})
