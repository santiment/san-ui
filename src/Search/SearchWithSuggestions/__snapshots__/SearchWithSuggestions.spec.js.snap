// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchWithSuggestions should render Bigbom as a child 1`] = `
<SearchWithSuggestions
  className=""
  data={
    Array [
      Object {
        "items": Array [
          "Bibox Token",
          "Bigbom",
          "Binance Coin",
          "BioCoin",
          "BitBay",
          "bitcoin",
        ],
        "predicate": [Function],
        "suggestionContent": [Function],
        "title": "Assets",
      },
    ]
  }
  debounceTime={200}
  defaultValue=""
  dontResetStateAfterSelection={false}
  inputProps={Object {}}
  maxSuggestions={5}
  onSuggestionSelect={[Function]}
  onSuggestionsUpdate={[Function]}
  suggestionsProps={Object {}}
  value=""
  withMoreSuggestions={true}
>
  <Search
    className=""
    icon="search-small"
    onChange={[Function]}
    onFocus={[Function]}
    wrapperRef={
      Object {
        "current": <div
          class="wrapper wrapper"
        >
          <input
            class="input input input left"
            placeholder="Type to search..."
            type="text"
            value="Big"
          />
          <svg />
          <svg />
        </div>,
      }
    }
  >
    <InputWithIcon
      className="wrapper"
      icon="search-small"
      iconClassName="icon"
      iconPosition="left"
      inputClassName="input"
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Type to search..."
      value="Big"
      wrapperRef={
        Object {
          "current": <div
            class="wrapper wrapper"
          >
            <input
              class="input input input left"
              placeholder="Type to search..."
              type="text"
              value="Big"
            />
            <svg />
            <svg />
          </div>,
        }
      }
    >
      <div
        className="wrapper wrapper"
      >
        <Input
          className="input input left"
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Type to search..."
          value="Big"
        >
          <input
            className="input input input left"
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Type to search..."
            type="text"
            value="Big"
          />
        </Input>
        <Component
          className="icon icon left"
          type="search-small"
        >
          <svg />
        </Component>
        <Component
          className="clear"
          onClick={[Function]}
          type="close-small"
        >
          <svg />
        </Component>
      </div>
    </InputWithIcon>
  </Search>
</SearchWithSuggestions>
`;

exports[`SearchWithSuggestions should render correctly 1`] = `
<Search
  className=""
  icon="search-small"
  onChange={[Function]}
  onFocus={[Function]}
  wrapperRef={
    Object {
      "current": null,
    }
  }
/>
`;